lib/src/main/java/graphql/nadel/engine/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -258,26 +261,45 @@ private class Factory(
                 actorFieldDef = actorFieldDef,
                 actorInputValueDefs = hydrationArgs,
             ),
-            sourceFields = hydrationArgs.mapNotNull {
-                when (it.valueSource) {
-                    is NadelHydrationActorInputDef.ValueSource.ArgumentValue -> null
-                    is FieldResultValue -> it.valueSource.queryPathToField
-                    is NadelHydrationActorInputDef.ValueSource.StaticValue -> null
-                }
-            },
-            condition = getHydrationCondition(hydration),
+            sourceFields = getHydrationSourceFields(hydrationArgs, condition),
+            condition = condition,
         )
     }
 
+    private fun getHydrationSourceFields(
+        hydrationArgs: List<NadelHydrationActorInputDef>,
+        condition: NadelHydrationWhenCondition?,
+    ): List<NadelQueryPath> {
+        val sourceFieldsFromArgs = hydrationArgs.mapNotNull {
+            when (it.valueSource) {
+                is NadelHydrationActorInputDef.ValueSource.ArgumentValue -> null
+                is FieldResultValue -> it.valueSource.queryPathToField
+                is NadelHydrationActorInputDef.ValueSource.StaticValue -> null
+            }
+        }
+
+        if (condition != null) {
+            return sourceFieldsFromArgs + condition.fieldPath
+        }
+
+        return sourceFieldsFromArgs
+    }
+
     private fun getHydrationCondition(hydration: UnderlyingServiceHydration): NadelHydrationWhenCondition? {
         if (hydration.conditionalHydration == null) {
             return null
         }
         if (hydration.conditionalHydration.predicate.equals != null) {
-            return NadelHydrationWhenCondition.ResultEquals(
-                fieldPath = NadelQueryPath(hydration.conditionalHydration.sourceField),
-                value = hydration.conditionalHydration.predicate.equals
-            )
+            when (val expectedValue = hydration.conditionalHydration.predicate.equals) {
+                is BigInteger -> return NadelHydrationWhenCondition.LongResultEquals(
+                    fieldPath = NadelQueryPath(hydration.conditionalHydration.sourceField),
+                    value = expectedValue.longValueExact(),
+                )
+                is String -> return NadelHydrationWhenCondition.StringResultEquals(
+                    fieldPath = NadelQueryPath(hydration.conditionalHydration.sourceField),
+                    value = expectedValue
+                )
```

Comment:
I think we should add an `else` step here and throw an error on unexpected type.